### PR TITLE
cli: Ignore `immutable()` in `jj git push` conflict/desc/user checks

### LINKED ISSUE #3029: FR: allow pushing commits with an empty author
**Is your feature request related to a problem? Please describe.**

context: https://discord.com/channels/968932220549103686/968932220549103689/1206485113949388860

context explained:
1. I am using `jj` with nixpkgs.
2. I rebased my custom `cmpkgs` branch on `nixos-unstable@nixos` (aka https://github.com/nixos/nixpkgs -> nixos-unstable branch)
3. I tried to push
4. I get:

```
Error: Won't push commit c2e3bdf541bc since it has no author and/or committer set
```

This stems from this commit: https://github.com/NixOS/nixpkgs/commit/c2e3bdf541bc8a2a5a348c1b030815aeef8b2723
from this PR:  https://github.com/NixOS/nixpkgs/pull/284580

**Describe the solution you'd like**

A flag to ignore this safety/sanity check and allow me to push anyway.

**Describe alternatives you've considered**
n/a seems unavoidable, I can't control upstream.

**Additional context**

**workaround**

```shell
> # this assumes you have named jj branch and 'origin' git branch the same
> b=cmpkgs # branch name
> # make sure you don't have [ui] color=always
> git -C .jj/repo/store/git push origin $(jj log -r ${b} -T commit_id --no-graph):$b -f
```

### PR DESCRIPTION
**Pull request by mlcui-corp**
Fixes #3029.

# Checklist

If applicable:
- [X] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [X] I have added tests to cover my changes

### CODE REVIEW

File: cli/src/commands/git/push.rs [hunk 1]
```diff
@@ -275,9 +276,16 @@ pub fn cmd_git_push(
...
         old_heads.push(repo.store().root_commit_id().clone());
     }
-    for commit in revset::walk_revs(repo.as_ref(), &new_heads, &old_heads)?
-        .iter()
-        .commits(repo.store())
+    // (old_heads | immutable())..new_heads
+    let commits_to_push = RevsetExpression::commits(old_heads)
+        .union(&revset_util::parse_immutable_expression(
+            &tx.base_workspace_helper().revset_parse_context(),
+        )?)
```
Comment (yuja) [inline cli/src/commands/git/push.rs:283]: nit: might be better to extract `parse_immutable_heads_expression()` that doesn't call `.ancestors()`.

File: cli/tests/test_git_push.rs [hunk 2]
```diff
@@ -835,6 +835,28 @@ fn test_git_push_no_description() {
     );
 }
 
+#[test]
+fn test_git_push_no_description_in_immutable() {
+    let (test_env, workspace_root) = set_up();
+    test_env.jj_cmd_ok(&workspace_root, &["branch", "create", "imm"]);
+    test_env.jj_cmd_ok(&workspace_root, &["describe", "-m="]);
+    test_env.add_config(r#"revset-aliases."immutable_heads()" = "imm""#);
```
Comment (yuja) [inline cli/tests/test_git_push.rs:843]: nit: for sanity check, can you add `git push --dry-run` without `immutable_heads() = imm`?

**Review by yuja — APPROVED** [review]
The idea sounds good to me.

File: cli/src/commands/git/push.rs (hunk 1 above)
Comment (mlcui-corp) [reply cli/src/commands/git/push.rs:280]: Done.

File: cli/tests/test_git_push.rs (hunk 2 above)
Comment (mlcui-corp) [reply cli/tests/test_git_push.rs]: Done.

**Comment by mlcui-corp** [comment]
PTAL, especially with the new changes in `cli/src/revset_util.rs`.

**Review by yuja — APPROVED** [review]
Thanks.

**Pull request merged**